packages/mui-base/src/Collapsible/Content/useCollapsibleContent.ts

```diff
@@ -0,0 +1,309 @@
+      process.env.NODE_ENV === 'test'
+        ? supportsOnBeforeMatch
+        : supportsCssContentVisibility && supportsOnBeforeMatch;
+  }
+  return cachedSupportsHiddenUntilFound;
+}
+
+/**
+ *
+ * Demos:
+ *
+ * - [Collapsible](https://mui.com/base-ui/react-collapsible/#hooks)
+ *
+ * API:
+ *
+ * - [useCollapsibleContent API](https://mui.com/base-ui/react-collapsible/hooks-api/#use-collapsible-content)
+ */
+export function useCollapsibleContent(
+  parameters: useCollapsibleContent.Parameters,
+): useCollapsibleContent.ReturnValue {
+  const {
+    animated = false,
+    htmlHidden = 'hidden',
+    id: idParam,
+    open,
+    mounted: contextMounted,
+    ref,
+    setContentId,
+    setMounted: setContextMounted,
+    setOpen,
+  } = parameters;
+
+  const id = useId(idParam);
+
+  const contentRef = React.useRef<HTMLElement | null>(null);
+
+  const [height, setHeight] = React.useState(0);
+
+  const latestAnimationNameRef = React.useRef<string | undefined>('none');
+  const originalTransitionDurationStyleRef = React.useRef<string | undefined>();
+
+  const isTransitioningRef = React.useRef(false);
+
+  useEnhancedEffect(() => {
+    setContentId(id);
+    return () => {
+      setContentId(undefined);
+    };
+  }, [id, setContentId]);
+
+  const handleContentRef = useEventCallback((element: HTMLElement) => {
+    if (!element) {
+      return;
+    }
+
+    contentRef.current = element;
+
+    const computedStyles = getComputedStyles(element);
+
+    latestAnimationNameRef.current = computedStyles.animationName ?? 'none';
+    originalTransitionDurationStyleRef.current = element.style.transitionDuration;
+  });
+
+  const mergedRef = useForkRef(ref, handleContentRef);
+
+  const runOnceAnimationsFinish = useAnimationsFinished(contentRef);
+
+  const isOpen = animated ? open || contextMounted : open;
+
+  const isInitialOpenRef = React.useRef(isOpen);
+
+  const isBeforeMatchRef = React.useRef(false);
+
+  useEnhancedEffect(() => {
+    const { current: element } = contentRef;
+
+    let frame1 = -1;
+    let frame2 = -1;
+
+    if (element) {
+      const computedStyles = getComputedStyles(element);
+      const currentAnimationName = computedStyles.animationName;
+      const isBeforeMatch = isBeforeMatchRef.current;
+      const isInitiallyOpen = isInitialOpenRef.current;
+      const isTransitioning = isTransitioningRef.current;
+      const originalAnimationName =
+        element.style.animationName === 'none' ? '' : element.style.animationName;
+      const originalTransitionDuration = originalTransitionDurationStyleRef.current;
+      const shouldCancelAnimation = isBeforeMatch || isInitiallyOpen;
+
+      element.style.animationName = 'none';
+
+      const rect = element.getBoundingClientRect();
+
+      if (!isTransitioning || !(open || contextMounted)) {
+        setHeight(rect.height);
+      }
+
+      element.style.animationName = shouldCancelAnimation ? 'none' : originalAnimationName;
```

Comment:
I'm wondering if we need to do all this on our side. Have you considered applying the `data-instant` attribute and instructing users to disable all motion when it's present? (perhaps this could also solve the issue with transitions)

Reply:
@michaldudak `data-instant` applies conditions that last for the lifetime of the component, which is different for the cases that are handled internally here, which only disables motion for one specific instance/occurrence of the `open` state changing:
- When the component is initially open, on first mount/load, motion is disabled so it's just appears fully open, but animations run normally after that
- When the component uses `hidden="until-found"` and is opened by in-page-search, it will open instantly, but animations run normally afterwards

I don't think these behaviors should be override-able for animated Collapsibles as it obviously results in poorer UX

Also this keeps the component easier to style, without having to add another animation state in addition to the 4 standard ones!